File: test/check-theme-mixing.test.js
import { test } from "node:test";
import assert from "node:assert/strict";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { run } from "../src/tools/check-theme-mixing.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const fixtures = path.join(__dirname, "fixtures");

test("flags a project that loads both Aura and Lumo", () => {
  const result = run({ positionals: ["mixed"], cwd: fixtures });
  assert.equal(result.ok, false);
  assert.deepEqual(result.themesLoaded.sort(), ["aura", "lumo"]);
  const error = result.findings.find((f) => f.code === "MULTIPLE_BASE_THEMES");
  assert.ok(error, "expected MULTIPLE_BASE_THEMES finding");
  assert.equal(error.level, "error");
});

test("passes a project that loads only one theme", () => {
  const result = run({ positionals: ["clean"], cwd: fixtures });
  assert.equal(result.ok, true);
  assert.deepEqual(result.themesLoaded, ["aura"]);
  assert.equal(result.findings.length, 0);
});

test("warns on mismatched theme tokens without failing the build", () => {
  const result = run({ positionals: ["mismatch"], cwd: fixtures });
  assert.equal(result.ok, true); // warning only, no error
  const warn = result.findings.find((f) => f.code === "MISMATCHED_THEME_TOKENS");
  assert.ok(warn, "expected MISMATCHED_THEME_TOKENS finding");
  assert.equal(warn.level, "warning");
});

test("flags LumoUtility usage while the Aura theme is loaded", () => {
  const result = run({ positionals: ["lumo-utility"], cwd: fixtures });
  assert.equal(result.ok, false);
  const error = result.findings.find(
    (f) => f.code === "LUMO_UTILITY_WITHOUT_LUMO_THEME"
  );
  assert.ok(error, "expected LUMO_UTILITY_WITHOUT_LUMO_THEME finding");
  assert.equal(error.level, "error");
  assert.equal(error.confidence, "high");
});

test("does not flag LumoUtility when the Lumo theme is loaded", () => {
  const result = run({ positionals: ["lumo-utility-ok"], cwd: fixtures });
  const error = result.findings.find(
    (f) => f.code === "LUMO_UTILITY_WITHOUT_LUMO_THEME"
  );
  assert.equal(error, undefined, "LumoUtility is valid under the Lumo theme");
});

test("reports a usage error for a missing directory", () => {
  const result = run({ positionals: ["does-not-exist"], cwd: fixtures });
  assert.ok(result.usageError);
});

File: go/internal/tools/checkthememixing_test.go
package tools

import (
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	"github.com/vaadin/agent-tools/internal/lib"
	"github.com/vaadin/agent-tools/internal/tool"
)

// fixturesDir resolves the shared test/fixtures directory at the repo root,
// reused verbatim from the JavaScript test suite (../../../test/fixtures
// relative to this file: go/internal/tools -> repo root -> test/fixtures).
func fixturesDir(t *testing.T) string {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("cannot locate test file path")
	}
	repoRoot := filepath.Join(filepath.Dir(file), "..", "..", "..")
	return filepath.Join(repoRoot, "test", "fixtures")
}

func run(t *testing.T, fixture string) themeMixingReport {
	t.Helper()
	return analyzeThemeMixing(tool.Args{Positionals: []string{fixture}, Cwd: fixturesDir(t)})
}

func findByCode(findings []lib.Finding, code string) *lib.Finding {
	for i := range findings {
		if findings[i].Code == code {
			return &findings[i]
		}
	}
	return nil
}

func TestFlagsBothAuraAndLumo(t *testing.T) {
	r := run(t, "mixed")
	if r.OK {
		t.Fatal("expected ok=false")
	}
	if got := strings.Join(r.ThemesLoaded, ","); got != "aura,lumo" {
		t.Fatalf("themesLoaded = %q, want aura,lumo", got)
	}
	f := findByCode(r.Findings, "MULTIPLE_BASE_THEMES")
	if f == nil {
		t.Fatal("expected MULTIPLE_BASE_THEMES finding")
	}
	if f.Level != "error" {
		t.Fatalf("level = %q, want error", f.Level)
	}
}

func TestDetectsFullyQualifiedAnnotations(t *testing.T) {
	r := run(t, "qualified")
	if r.OK {
		t.Fatal("expected ok=false for fully qualified Aura + Lumo")
	}
	if got := strings.Join(r.ThemesLoaded, ","); got != "aura,lumo" {
		t.Fatalf("themesLoaded = %q, want aura,lumo", got)
	}
	if findByCode(r.Findings, "MULTIPLE_BASE_THEMES") == nil {
		t.Fatal("expected MULTIPLE_BASE_THEMES finding from qualified annotations")
	}
}

func TestPassesSingleTheme(t *testing.T) {
	r := run(t, "clean")
	if !r.OK {
		t.Fatal("expected ok=true")
	}
	if got := strings.Join(r.ThemesLoaded, ","); got != "aura" {
		t.Fatalf("themesLoaded = %q, want aura", got)
	}
	if len(r.Findings) != 0 {
		t.Fatalf("expected no findings, got %d", len(r.Findings))
	}
}

func TestWarnsOnMismatchedTokens(t *testing.T) {
	r := run(t, "mismatch")
	if !r.OK {
		t.Fatal("expected ok=true (warning only)")
	}
	f := findByCode(r.Findings, "MISMATCHED_THEME_TOKENS")
	if f == nil {
		t.Fatal("expected MISMATCHED_THEME_TOKENS finding")
	}
	if f.Level != "warning" {
		t.Fatalf("level = %q, want warning", f.Level)
	}
}

func TestFlagsLumoUtilityUnderAura(t *testing.T) {
	r := run(t, "lumo-utility")
	if r.OK {
		t.Fatal("expected ok=false")
	}
	f := findByCode(r.Findings, "LUMO_UTILITY_WITHOUT_LUMO_THEME")
	if f == nil {
		t.Fatal("expected LUMO_UTILITY_WITHOUT_LUMO_THEME finding")
	}
	if f.Level != "error" {
		t.Fatalf("level = %q, want error", f.Level)
	}
	if f.Confidence != "high" {
		t.Fatalf("confidence = %q, want high", f.Confidence)
	}
}

func TestDoesNotFlagLumoUtilityUnderLumo(t *testing.T) {
	r := run(t, "lumo-utility-ok")
	if f := findByCode(r.Findings, "LUMO_UTILITY_WITHOUT_LUMO_THEME"); f != nil {
		t.Fatal("LumoUtility is valid under the Lumo theme; should not be flagged")
	}
}

func TestNoOpWhenThemeIndeterminate(t *testing.T) {
	r := run(t, "indeterminate")
	if !r.OK {
		t.Fatal("expected ok=true (correctness out of scope here)")
	}
	if len(r.ThemesLoaded) != 0 {
		t.Fatalf("themesLoaded = %v, want empty", r.ThemesLoaded)
	}
	f := findByCode(r.Findings, "THEME_INDETERMINATE")
	if f == nil {
		t.Fatal("expected THEME_INDETERMINATE info finding")
	}
	if f.Level != "info" {
		t.Fatalf("level = %q, want info", f.Level)
	}
	for _, x := range r.Findings {
		if x.Level == "error" || x.Level == "warning" {
			t.Fatalf("unexpected %s finding when theme is unknown: %s", x.Level, x.Code)
		}
	}
}

func TestUsageErrorForMissingDir(t *testing.T) {
	r := run(t, "does-not-exist")
	if r.UsageError == "" {
		t.Fatal("expected a usage error for a missing directory")
	}
}
